### ISSUE TITLE
The filter out data is wrong when use multiple jsgrid in tabs

### ISSUE THREAD

**Issue by jufei** [tabalinas/jsgrid#219]
hi:
   I tried to use jquery tabs UI and in each tab, I wish to show a grid that connects to a table in database.

When I firstly access the URL, yes, everything is OK and all data are shown correctly in each tab each grid. But when I do some filter in the first grid, I found that the filter out data is belong to the last grid, not for the first grid. Same result is the second grid and the third grid.

I have checked the loadData function for the grid, yes I have sent the right data to the grep function:

```js
                        loadData: function(filter) {
                            console.log(datacenter.database.clients.length);
                            return $.grep(datacenter.database.clients, function(client) {
```

or

```js
loadData: function(filter) {
                            console.log(datacenter[sheetname].clients.length);
                            return $.grep(datacenter[sheetname].clients, function(client) {
```

I think for each grid, the first parameter of grep function should be the datasource for the grid. But it is wrong.

I don't know why.

Attach is my html file with the javascript in it.

Thanks in advance.
[database.zip](https://github.com/tabalinas/jsgrid/files/162652/database.zip)

**Comment by tabalinas** [comment]
The issue is not clear to me. Please prepare a jsfiddle (e.g. modifying this one https://jsfiddle.net/tabalinas/h51z9syt/) to demonstrate the issue.

> I think for each grid, the first parameter of grep function should be the datasource for the grid. But it is wrong.

This is not clear. What do you mean?

**Comment by jufei** [comment]
> 

> I think for each grid, the first parameter of grep function should be the datasource for the grid. But it is wrong.

I mean: how the grid show data? I think it will execute the function for loaddata, right?  Currently in my loaddata function, i used: return $.grep(datacenter[sheetname].clients, function(client)....; so I guessthe loaddata function will go through the datasource and filter out records with the filter function, so the data source for the grid should be datacenter[sheetname].clients, right?

**Comment by jufei** [comment]
I added some code to my html to debug it and I found:
1. At the beginning, my 5 grids load data correctly;
2. All my grid has filtering: true, that means each grid will have a filter row;
3. When I tried to filter data in the 1st grid by selecting a value from a listbox in the filter row, according to the console log, the loaddata function of the 5th grid was triggered. That is why the data of 5th grid was shown in 1st grid.

**Comment by jufei** [comment]
I use debug to trace into the js file and found the error happens in jsgrid.core.js line 865

>             return $.when(controller[method](param))
                .done($.proxy(doneCallback, this))
                .fail($.proxy(this._errorHandler, this))
                .always($.proxy(this._hideLoading, this));

I found here the controller is the controller of 5th grid, not the controller of 1st grid, that is why the data shown error.

**Comment by tabalinas** [comment]
Still your report is not clear. If you still need help with the case, please provide a working fiddle demonstrating the issue.
Or at least provide an archive with working example (attached `database.zip` requires a lot of external not included scripts to make it work).

**Comment by jufei** [comment]
Thanks for your response first.

Yes, my html file contains some external files and the data is gotten by ajax, so it will not work for you. And yes, it is difficult to describe it clearly. 

This morning, I go on debug this issue, and found that the controller itself is correct, but the ownership for the jsgrid and the controller is wrong, all my 5 jsgrid's controller become the controller of the 5th grid.

I am a newer for javascript and not know too much skills for debugging it. But I continue work for it.

Anyway, Thanks.

**Comment by tabalinas** [comment]
Ok, feel free to post a comment if you have questions or will find an issue.

**Comment by jufei** [comment]
你们那个关闭信号，db是多少啊？


BR.
Jufei.
Location: 02A08, Tel: 13606607936

**Comment by jufei** [comment]
Hi:
          Let’t me try to explain it again.

          I have a html page, in the page, I used 5 jsgrids, each jsgrid has its own controller to control its data.

          The 5 jsgrids appeared successfully and all data are right.

          But when I try to do some filter operation in the 1 grid, I found that the data of the 1st jsgrid was replaced by the data of 5th jsgrid.
          Also I found the 2nd ,3rd ,4th jsgrid’s data were replaced by the data of 5th jsgrid.


          Now I am sure that the grid’s controller has been changed and the 5th grid’s controller was attached to each jsgrid.

Would you please tell me which part of the jsgrid.core.js is the action to attach a controller to a jsGrid? I only find this:

_initController: function() {
            this._controller = getOrApply(this.controller, this);
        },

Thanks.

BR.
Jufei.
Location: 02A08, Tel: 13606607936

**Comment by jufei** [comment]
Hi:
          So for me, the directly way is to debug when or where the controller of the 1st grid was replaced.

          But as I said I am a beginner of javascript, and I don’t know how to trace the 1st grid with console.log. en.

Thanks.

**Comment by tabalinas** [comment]
Try to start with something simple in jsFiddle (https://jsfiddle.net/tabalinas/h51z9syt/).
Add 2 tabs with a grid on each tab. Replace ajax calls with return of static items, or just use `setTimeout()` to imitate asynchronous request.
When you'll have a simple fiddle without cluttering stuff, you will whether find an issue yourself, or, if not, will provide this example here, and so we will be able to help you. 
Current explanation without an example won't help.

**Comment by jufei** [comment]
It’s OK now. The problem is caused because I used the jsgrid.core.js is a old version. I got it from your demo webpage directly. And it seems is very old.

Anyway, thanks.

BR.
Jufei.
Location: 02A08, Tel: 13606607936

**Comment by tabalinas** [comment]
Glad that you figured it out.
But I'm not sure about "old scripts on the site". What do you mean? I've checked scripts on the http://js-grid.com and it seems to be the latest. Could you please tell me where scripts should be updated?

**Comment by jufei** [comment]
I could not remember where I got all the js files, but in my jsgrid.core.cs,  the init function is like this:
        _init: function(config) {
            $.extend(true, this, config);
            this._initLoadStrategy();
            this._initController();
            this._initFields();
            this._attachWindowLoadResize();
            this._attachWindowResizeCallback();
        },


Also the jsgrid.field.select.js, it has no filter_value, insert_value and edit_value function, it causes a problem:  when I use a select which use valuefield is string, it will cause error. I modified it myself.


BR.
Jufei.
Location: 02A08, Tel: 13606607936

**Comment by michaeldeongreen** [comment]
I feel like I came across an issue that was similar to this ie multiple tabs with mutliple jsgrids on each tab and certain events being called were from the wrong grid.

Basically, I have:

3 Tabs
2 jsGrids per Tab

On each tab, one of my grids has an Add Button (right grid) and when this button is pressed, it will add the item to the grid on the left.

**Below is a code snippet:**
`	$("somegrid").jsGrid({
		......
            fields: [
                { name: "add", type: "addButtonField", width: 35, title: "", filtering: false }
            ]
	});

    self.configureAddButtonField = function () {
        var field = function (config) {
            jsGrid.Field.call(this, config);
        }
        field.prototype = new jsGrid.Field({
            itemTemplate: function (value, item) {
                return $("<button>").text("Add")
                    	.on("click", function () {
                    	    self.parentViewModel.selectedCategoriesGrid.addNewCategory(item);
                    	}).addClass("btn btn-info");
            }
        });
        jsGrid.fields.addButtonField = field;
    }`

What I noticed is that the onClick event would call the onClick event of the last grid added to the html page and not the onClick of the Add Button of the actual grid that you were clicking.

1) When I would press the Add Button on say the Grid on the first Tab, it would add the item to the Grid on the 3rd Tab

2) When I removed the 3rd Tab, the 1st Tab's grid would add the item to the 2nd Tab's grid, which was how I identified the pattern.

The fix:

1) I had to make sure each grid that has an Add Button had a different name for the "type" field on the grid ie addButtonField1, addButtonField2, addButtonField3 etc etc.
